metadata-ingestion/src/datahub/ingestion/source/metabase.py

```diff
@@ -0,0 +1,407 @@
+
+        return custom_properties
+
+    def get_datasource_urn(self, card_details):
+        platform, database_name = self.get_datasource_from_id(
+            card_details["database_id"]
+        )
+        query_type = card_details.get("dataset_query", {}).get("type", {})
+        if query_type == "query":
+            source_table_id = (
+                card_details.get("dataset_query", {})
+                .get("query", {})
+                .get("source-table", {})
+            )
+            schema_name, table_name = self.get_source_table_from_id(source_table_id)
+            if table_name:
+                platform_urn = f"urn:li:dataPlatform:{platform}"
+                dataset_urn = (
+                    f"urn:li:dataset:("
+                    f"{platform_urn},{database_name + '.' if database_name else ''}"
+                    f"{schema_name + '.' if schema_name else ''}"
+                    f"{table_name},{self.config.env})"
+                )
+                return [dataset_urn]
+        else:
+            self.report.report_warning(
+                key=f"metabase-card-{card_details['id']}",
+                reason=f"Cannot create datasource urn from query type: {query_type}",
+            )
+
+        return None
+
+    @lru_cache(maxsize=None)
+    def get_source_table_from_id(self, table_id):
+        dataset_response = self.session.get(
+            f"{self.config.connect_uri}/api/table/{table_id}"
+        ).json()
+
+        schema = dataset_response.get("schema", "")
+        name = dataset_response.get("name", "")
+
+        return schema, name
+
+    @lru_cache(maxsize=None)
+    def get_datasource_from_id(self, datasource_id):
+        dataset_response = self.session.get(
+            f"{self.config.connect_uri}/api/database/{datasource_id}"
+        ).json()
+
+        # Map engine names to what datahub expects in
+        # https://github.com/linkedin/datahub/blob/master/metadata-service/war/src/main/resources/boot/data_platforms.json
+        engine = dataset_response["engine"]
+        platform = engine
+
+        engine_mapping = {
+            "sparksql": "spark",
+            "mongo": "mongodb",
+            "presto-jdbc": "presto",
+            "sqlserver": "mssql",
+            "bigquery-cloud-sdk": "bigquery",
+        }
+        if engine in engine_mapping:
+            platform = engine_mapping[dataset_response["engine"]]
+
+        field_for_dbname_mapping = {
+            "postgres": "dbname",
+            "sparksql": "dbname",
+            "mongo": "dbname",
+            "redshift": "db",
+            "snowflake": "db",
+            "presto-jdbc": "catalog",
+            "presto": "catalog",
+            "mysql": "dbname",
+            "sqlserver": "db",
+        }
+
+        dbname = (
+            dataset_response.get("details", {}).get(
+                field_for_dbname_mapping[dataset_response["engine"]]
+            )
+            if engine in field_for_dbname_mapping
+            else None
```

Comment:
Are there any reasonable fallbacks here? Does it make sense to just try "db", "dbname", and "catalog" if the engine is unknown? 

If not, should we use `self.report.report_warning` to alert the end-user of an issue?

Reply:
Similar question for the `platform` and `engine_mapping` above

Reply:
We now report warning when we are unable to resolve platform or dbname. see [commit](https://github.com/linkedin/datahub/pull/3602/commits/ea905000fa8390af5e20ec3f2abea7b944a9cab9)